### PR TITLE
issue #632 _typeFilter support for Bulkdata Export jobs

### PR DESCRIPTION
**Pull request by albertwang-ibm**

### CODE REVIEW

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java [hunk 1]
```diff
@@ -213,6 +213,10 @@ public String submit(Instant since, List<String> types, Map<String, String> prop
             builder.fhirSearchFromDate("1970-01-01");
         }
 
+        if (properties.get(BulkDataConstants.PARAM_TYPE_FILTER) != null) {
+            builder.fhirTypeFilters(properties.get(BulkDataConstants.PARAM_TYPE_FILTER));
+        }
```
Comment (prb112) [inline fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java]: Totally get why it was missed!

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java [hunk 2]
```diff
@@ -144,37 +159,48 @@ private void fillChunkDataBuffer(List<Member> patientRefs) throws Exception {
...
+                        FHIRTransactionHelper txn = new FHIRTransactionHelper(fhirPersistence.getTransaction());
+                        txn.enroll();
+                        FHIRPersistenceContext persistenceContext = FHIRPersistenceContextFactory.createPersistenceContext(null, searchContext);
+                        List<Resource> resources2 = fhirPersistence.search(persistenceContext, resourceType).getResource();
+                        txn.unenroll();
+                        compartmentPageNum++;
+
+                        for (Resource res2 : resources2) {
+                            if (res2 == null) {
+                                continue;
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java]: why would it be null?

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java [hunk 3]
```diff
@@ -144,37 +159,48 @@ private void fillChunkDataBuffer(List<Member> patientRefs) throws Exception {
...
+                            }
+                            try {
+                                FHIRGenerator.generator(Format.JSON).generate(res2, chunkData.getBufferStream());
+                                chunkData.getBufferStream().write(Constants.NDJSON_LINESEPERATOR);
+                                resSubTotal++;
+                            } catch (FHIRGeneratorException e) {
+                                logger.log(Level.WARNING, "fillChunkDataBuffer: Error while writing resources with id '"
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java]: so this ends in a warning... does the whole job fail?

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 4]
```diff
@@ -110,18 +121,23 @@ public ChunkReader() {
...
-                    continue;
-                }
-                Patient patient = (Patient) res;
+            do {
                 Map<String, List<String>> queryParameters = new HashMap<>();
+                // Add the search parameters from the current typeFilter for current resource type.
+                if (searchParametersForResoureTypes.get(resourceType) != null) {
+                    queryParameters.putAll(searchParametersForResoureTypes.get(resourceType).get(indexOfCurrentTypeFilter));
+                }
                 List<String> searchCreteria = new ArrayList<>();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]:
```suggestion
                List<String> searchCriteria = new ArrayList<>();
```

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java [hunk 5]
```diff
@@ -194,11 +212,16 @@ public Object readItem() throws Exception {
...
         FHIRSearchContext searchContext;
         FHIRPersistenceContext persistenceContext;
         Map<String, List<String>> queryParameters = new HashMap<>();
 
+        // Add the search parameters from the current typeFilter for current resource type.
+        if (searchParametersForResoureTypes.get(resourceType) != null) {
+            queryParameters.putAll(searchParametersForResoureTypes.get(resourceType).get(indexOfCurrentTypeFilter));
+        }
+
         List<String> searchCreteria = new ArrayList<String>();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]:
```suggestion
        List<String> searchCriteria = new ArrayList<String>();
```

**Comment by prb112** [comment]
Small Comments above

**Comment by albertwang-ibm** [comment]
> Is this still WIP? Approving if it's done

I'm trying to make it better, like removing some duplicated codes, avoid exporting the same resource multiple times and with more tests ... will remove WIP when finished, thanks!!

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java (hunk 2 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java]: just for code safety, routine check. :)

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java (hunk 3 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java]: just logging and continue.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 4 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: ha, good catch, let me refactor in my eclipse.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java (hunk 5 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 4 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 6]
```diff
@@ -48,34 +50,43 @@
...
-    private FHIRPersistence fhirPersistence;
-    private List<String> resourceTypes;
+    protected FHIRPersistence fhirPersistence;
+    protected List<String> resourceTypes;
+
+    // Search parameters for resource types gotten from fhir.typeFilters job parameter.
+    Map<Class<? extends Resource>, List<Map<String, List<String>>>> searchParametersForResoureTypes = null;
+    // Used to prevent the same resource from being exported multiple times when multiple _typeFilter for the same
+    // resource type are used, which leads to multiple search requests which can have overlaps of resources.
+    HashSet<String> loadedResourceIds = new HashSet<>();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: Is there a specific HashSet method you are using? or best to make use of Set

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 7]
```diff
@@ -108,73 +126,88 @@ public ChunkReader() {
...
+
+                        for (Resource res : resources) {
+                            if (res == null || (isDoDuplicationCheck && loadedResourceIds.contains(res.getId()))) {
+                                continue;
+                            }
+                            try {
+                                FHIRGenerator.generator(Format.JSON).generate(res, chunkData.getBufferStream());
+                                chunkData.getBufferStream().write(Constants.NDJSON_LINESEPERATOR);
+                                resSubTotal++;
+                                if (isDoDuplicationCheck) {
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: is there some documentation on this behavior?  Maybe add a comment here, just to explain when another developer sees this. (maybe on line 177) 
you can combine this logic with line 177-178

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 8]
```diff
@@ -108,73 +126,88 @@ public ChunkReader() {
...
+                                    loadedResourceIds.add(res.getId());
+                                }
+                            } catch (FHIRGeneratorException e) {
+                                if (res.getId() != null) {
+                                    logger.log(Level.WARNING, "fillChunkDataBuffer: Error while writing resources with id '"
+                                            + res.getId() + "'", e);
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: should this be part of an operational outcome?

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 9]
```diff
@@ -108,73 +126,88 @@ public ChunkReader() {
...
+                                } else {
+                                    logger.log(Level.WARNING,
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: same comment as prior one - operationaloutcome / warning

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 10]
```diff
@@ -108,73 +126,88 @@ public ChunkReader() {
...
-                } while (searchContext.getLastPageNumber() >= compartmentPageNum);
-            }
+                    } while (searchContext.getLastPageNumber() >= compartmentPageNum);
+                }
+
+                indexOfCurrentTypeFilter++;
+            } while (searchParametersForResoureTypes.get(resourceType) != null && indexOfCurrentTypeFilter < searchParametersForResoureTypes.get(resourceType).size());
+
             chunkData.setCurrentPartResourceNum(chunkData.getCurrentPartResourceNum() + resSubTotal);
             logger.fine("fillChunkDataBuffer: Processed resources - " + resSubTotal + "; Bufferred data size - "
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: this didn't occur to me earlier, maybe wrap this in isLogLevel

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 11]
```diff
@@ -205,27 +227,11 @@ public Object readItem() throws Exception {
...
-            logger.fine("readItem: Set DatastoreId to default!");
-        }
-        if (fhirSearchPageSize != null) {
-            try {
-                pageSize = Integer.parseInt(fhirSearchPageSize);
-                logger.fine("readItem: Set page size to " + pageSize + ".");
-            } catch (Exception e) {
-                logger.warning("readItem: Set page size to default(" + Constants.DEFAULT_SEARCH_PAGE_SIZE + ").");
+                isDoDuplicationCheck = false;
+                loadedResourceIds.clear();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: I know we had a similar pattern in this code before with clear. 
I see this and I think... are there unintended side effects if not properly managed.  

Maybe add a explanation of the behavior to the initial definition.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 12]
```diff
@@ -268,7 +273,11 @@ public Object readItem() throws Exception {
 
         if (resources != null) {
             logger.fine("readItem(" + resourceTypes.get(indexOfCurrentResourceType) + "): loaded patients number - " + resources.size());
-            fillChunkDataBuffer(resources);
+
+            List<String> patientIds = resources.stream().filter(item -> item.getId() != null).map(item -> item.getId()).collect(Collectors.toList());
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: Nice line of code.  Very handy.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java [hunk 13]
```diff
@@ -50,6 +52,17 @@
     int indexOfCurrentResourceType = 0;
     // Control the number of records to read in each "item".
     int pageSize = Constants.DEFAULT_SEARCH_PAGE_SIZE;
+    // Search parameters for resource types gotten from fhir.typeFilters job parameter.
+    Map<Class<? extends Resource>, List<Map<String, List<String>>>> searchParametersForResoureTypes = null;
+    int indexOfCurrentTypeFilter = 0;
+
+    // Used to prevent the same resource from being exported multiple times when multiple _typeFilter for the same
+    // resource type are used, which leads to multiple search requests which can have overlaps of resources.
+    HashSet<String> loadedResourceIds = new HashSet<>();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: same comment on HashSet

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java [hunk 14]
```diff
@@ -115,14 +132,17 @@ private void fillChunkDataBuffer(List<Resource> resources) throws Exception {
...
+                if (res == null || (isDoDuplicationCheck && loadedResourceIds.contains(res.getId()))) {
                     continue;
                 }
 
                 try {
                     FHIRGenerator.generator(Format.JSON).generate(res, chunkData.getBufferStream());
                     chunkData.getBufferStream().write(Constants.NDJSON_LINESEPERATOR);
                     resSubTotal++;
+                    if (isDoDuplicationCheck) {
+                        loadedResourceIds.add(res.getId());
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: sorry I misunderstood the prior isDoDuplicationCheck logic

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 7 above)
Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: please ignore

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java [hunk 15]
```diff
@@ -146,70 +166,59 @@ private void fillChunkDataBuffer(List<Resource> resources) throws Exception {
...
-        List<String> searchCreteria = new ArrayList<String>();
+        // Add the search parameters from the current typeFilter for current resource type.
+        if (searchParametersForResoureTypes.get(resourceType) != null) {
+            queryParameters.putAll(searchParametersForResoureTypes.get(resourceType).get(indexOfCurrentTypeFilter));
+            if (searchParametersForResoureTypes.get(resourceType).size() > 1) {
+                isDoDuplicationCheck = true;
+            }
+        }
+
+        List<String> searchCriteria = new ArrayList<String>();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: just a thought, maybe put this in a utility class to process the query parameters

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java [hunk 16]
```diff
@@ -218,15 +227,14 @@ public Object readItem() throws Exception {
         searchContext.setPageNumber(pageNum);
         List<Resource> resources = null;
         FHIRTransactionHelper txn = new FHIRTransactionHelper(fhirPersistence.getTransaction());
-        txn.begin();
+        txn.enroll();;
         persistenceContext = FHIRPersistenceContextFactory.createPersistenceContext(null, searchContext);
         resources = fhirPersistence.search(persistenceContext, resourceType).getResource();
-        txn.commit();
+        txn.unenroll();;
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: extra `;`
```suggestion
        txn.unenroll();
```

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java [hunk 17]
```diff
@@ -218,15 +227,14 @@ public Object readItem() throws Exception {
         searchContext.setPageNumber(pageNum);
         List<Resource> resources = null;
         FHIRTransactionHelper txn = new FHIRTransactionHelper(fhirPersistence.getTransaction());
-        txn.begin();
+        txn.enroll();;
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]:
```suggestion
        txn.enroll();
```

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 18]
```diff
@@ -108,73 +126,88 @@ public ChunkReader() {
...
+
+                for (String patientId : patientIds) {
+
+                    searchContext = SearchUtil.parseQueryParameters("Patient", patientId,
+                            ModelSupport.getResourceType(resourceTypes.get(indexOfCurrentResourceType)), queryParameters, true);
+                    do {
+                        searchContext.setPageSize(pageSize);
+                        searchContext.setPageNumber(compartmentPageNum);
+                        FHIRTransactionHelper txn = new FHIRTransactionHelper(fhirPersistence.getTransaction());
+                        txn.enroll();;
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]:
```suggestion
                        txn.enroll();
```

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 19]
```diff
@@ -108,73 +126,88 @@ public ChunkReader() {
...
+                        FHIRPersistenceContext persistenceContext = FHIRPersistenceContextFactory.createPersistenceContext(null, searchContext);
+                        List<Resource> resources = fhirPersistence.search(persistenceContext, resourceType).getResource();
+                        txn.unenroll();;
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]:
```suggestion
                        txn.unenroll();
```

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java [hunk 20]
```diff
@@ -768,6 +768,7 @@ protected synchronized Client getClient() throws Exception {
             // Add a hostname verifier if we're using an ssl transport.
             if (usingSSLTransport() && !isHostnameVerificationEnabled()) {
                 cb = cb.hostnameVerifier(new HostnameVerifier() {
+
```
Comment (prb112) [inline fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: Is there another change here?

Comment (albertwang-ibm) [reply fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: ah, sorry, this file is added by mistake...

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 19 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: good catch, my eclipse has rendering issue, I have to resize the editor window the show the correct contents sometime, seems I missed these, thanks! considering to upgrade my eclipse now, I'm still using 2019 June version ...

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java (hunk 20 above)
Comment (prb112) [reply fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: no problem - just confirming.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 8 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: export doesn't require operation outcome in fhir spec, and this is why I didn't add like what was did in import, so just logging ...

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 10 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: sounds good, let me change it.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 8 above)
Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: definitely something we should think about.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 6 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: changing the declaration to Set (interface) works, but what I really want is just a Hashset, not treeset, linkhashset etc, so I thought not need to use interface for the declaration.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 9 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: export doesn't require operation outcome in fhir spec, and this is why I didn't add like what was did in import, so just logging ...

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 11 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: make sense, let me add.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 12 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: yeah, sometime I like java8. :)

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java (hunk 13 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 11 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 10 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 6 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java (hunk 15 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: en... good thought! will consider in the next big change for export jobs - issue #779 partitionize the export jobs.

Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ChunkReader.java]: yeah - no worries - this was just a thought

**Pull request merged**